### ISSUE TITLE
Enhance Power Flow output on diagrams

### ISSUE THREAD

**Issue by ceraolo** [PowerGrids/PowerGrids#116]
**This ticket regards only the PF feature of the new PowerGrids.**

Now we have the possibility to see the PF results directly on the system diagrams. This is a very important and useful feature. Indeed, when doing PF studies, I think users will 90% of the time have a look at just the diagram with numerical result on it, and only rarely will read other results through the output variables.

I like to seee these results in PU. However, for this case, I think it is important to make a change to make it even more useful.

Now active and reactive powers are referred to the nominal machines they belong to, which in general are different from different machines in the system. For instance a 100 MVA generator can be connected to a 130 MVA transformer.
So, when we output the PF results on the diagram in PU, PU numbers are not directly comparable. 
Consider for instance bus33 in the picture below. the power coming from pv33 is 100 MW, which goes into the two transformers below it. In fact their nominal powers are 75 MVA, so they receive 0.6667*75= 50 MW each. This is not immediately evident from the numerical labels on the diagram, since they refer to different base units. 

The situation could be improved a lot if all active and reactive powers are shown in a unique base, e.g. 100 MVA for the whole grid.

My proposal is therefore to add to the SystemObject an option to be selected through a checkBox, whose name could be "Use unique power base for power display on diagrams".
The this checkbox is selected an input field should become active, allowing specifying the value of the unique power base (e.g. 100 MVA).

In that case the user would have a clear picture of the power flow just having a look at the diagram:

- they could select the UNom value for buses that to correspond to the adjacent transformer windings
- they could select a unique power base for the system e.g. 100 MVA. This will override local nominal values for the computation of the PQ active and reactive powers displayed on diagrams.

Is this possible with a reasonable effort?
Does anyone agree that this could be helpful?

![image](https://github.com/PowerGrids/PowerGrids/assets/7373030/4b2116ba-08fa-4c98-bb9d-c3db31b540d8)

**Comment by casella** [comment]
To be honest, I'm not sure I see the advantage of that over just looking at the MW outputs. If you set the base power to 100 MVA, the difference would just be that the decimal point is moved by two places... That said, if that is what power engineers are used to, so be it.

The whole iPSL library is based on this concept, and I found it quite confusing. But if we keep it at the level of displaying (not writing models), it's fine for me.

BTW, I guess this only makes sense for power, not for voltage, so it would be SNom, not UNom.

The implementation of this feature should be fairly straightforward. However, I would keep adding new features that are not backwards-incompatible for the 2.1.0 version, I think now we should focus on releasing what we have ASAP.

**Comment by ceraolo** [comment]
> To be honest, I'm not sure I see the advantage of that over just looking at the MW outputs. If you set the base power to 100 MVA, the difference would just be that the decimal point is moved by two places... That said, if that is what power engineers are used to, so be it.

That's right. But numbers on diagrams make them cluttered, and writing 1.234  (pu of 1 GW) is more compact than 1234000 (in W, i.e., pu of one watt). I think four significant digits are sufficient for numerical data on diagrams.
I agree it is a minor issue,  though.

**Comment by ceraolo** [comment]
> To be honest, I'm not sure I see the advantage of that over just looking at the MW outputs. If you set the base power to 100 MVA, the difference would just be that the decimal point is moved by two places... That said, if that is what power engineers are used to, so be it.

Quite true. 
This ticket could be useful, however, to have powers that sum at nodes (the sum of powers flowing into them being zero) and still bus voltages in pu.
But I agree that it has low priority, if some. We will see what happens when people start (hopefully) using the library regularly.

**Comment by AndreaBartolini** [comment]
currently it is possible:
- to visualize all quantities in p.u.
- to visualize all quantities in SI

The user can select the visualization mode by setting the following flags in the `system` object:
- `showDataOnDiagramsPu`
- `showDataOnDiagramsSI`

These flags are mutually exclusive, when one is set the other becomes unavailable (disabled). If both are not set the visualization is turned off.

**Comment by AndreaBartolini** [comment]
@ceraolo  we can add, if it is useful, a third possibility: power in SI, voltage and phase in p.u.

**Comment by AndreaBartolini** [comment]
@ceraolo we can add also the possibility to define in the `system` object a unique `SBaseDisp` and a unique `UBaseDisp` to be used as common base for the visualization.

Let me know what are the best solutions to be implemented

**Comment by ceraolo** [comment]
> currently it is possible:
> 
> * to visualize all quantities in p.u.
> * to visualize all quantities in SI
> 
> The user can select the visualization mode by setting the following flags in the `system` object:
> 
> * `showDataOnDiagramsPu`
> * `showDataOnDiagramsSI`
> 
> These flags are mutually exclusive, when one is set the other becomes unavailable (disabled). If both are not set the visualization is turned off.

Please consider this ticket's description. Saying "PU" is vague: which base?

**Comment by ceraolo** [comment]
> [@ceraolo](https://github.com/ceraolo) we can add, if it is useful, a third possibility: power in SI, voltage and phase in p.u.

Personally I would not use this option. It seems to me it adds confusion instead of clarity.

**Comment by ceraolo** [comment]
> [@ceraolo](https://github.com/ceraolo) we can add also the possibility to define in the `system` object a unique `SBaseDisp` and a unique `UBaseDisp` to be used as common base for the visualization.
> 
> Let me know what are the best solutions to be implemented

I have no much time to follow now all recent changes on Powergrids, that come years after I spent a lot of time on it. Sorry for this.

Let me just state a few points.

Even though researchers tend to forget, using PU is just using a base different from SI for components (and using different bases for different components). Researchers tend to say that "PU values are dimensionless", in contrast from SI ones, but this is incorrect: CGPM always clarifies that every numerical number is actually dimensionless since it is the ratio of a physical value and its unit of measure. 
_Based on this simple consideration the introduction of PU that I make to students, and the corresponding simplifications in the analysis, is 2-3 pages, instead of 15-20._ 

But I don't want to add excess entropy to this ticket: the text I write below can be easily understand by anyone, even those that believe that "PU values are dimensionless".

I think that the display can follow two lines of thought.

**1) display in an unique base for the whole system** 
the default base should be 1MVA - 1 kV, which is very effective for HV and larger voltage systems, since the displayed numbers, will usually have a few (not zero, not many) digits before the decimal point. Other bases could help users, especially when they simulate MV systems: for instance setting SBaseDisp=100kVA and UBaseDisp=100 V would help in this respect: numbers will  be in tens of a 100MVA base and tens of a 1 kV base.  

**2) display in different bases for individual components (PU)** 
In this case for each component we will see the PU values with reference to their bases. Obviously, power fluxes cannot be easily summed up, as discussed in this ticket's description. This option might help in csome cases, e.g. when we want to see whether components are working outside their nominal values. Maybe this option can be complemented by specific font (e.g. bold or italic or underlined or blinkinc) when the component's size (its MVAs, its kVs) is overcome.

Note that option 1 above coincides with "use PU with a unique base for the whole system"

**Comment by AndreaBartolini** [comment]
Let me understand... 
the point (as far as I understood) is that the current visualization in PU may refer to different bases because there is no a unique base in the library (each component uses a local base defined by the local UNom and the SNom), and this may be a bit confusing.

The idea to use a unique base for the visualization on the diagram (to be set in the `system` object) is in general good for me, the only caveat I see is that the values of the same variable in PU displayed on the trend may be different, because the trend uses the local base of the component.

If this difference between the bases used for the diagram and for the trends is not a problem I can easily implement the modification...
